packages/ckeditor5-html-support/src/datafilter.ts

```diff
@@ -805,129 +807,133 @@ export interface DataFilterRegisterEvent {
+				result.attributes.push( attributeName );
+			}
+		}
 
-	return viewAttributes;
+		return result;
+	}, {
+		attributes: [] as Array<string>,
+		classes: [] as Array<string>,
+		styles: [] as Array<string>
+	} );
 }
 
 /**
- * Consumes matched attributes.
- *
- * @returns Array with match information about found attributes.
+ * Prepares the GHS attribute value as an object with element attributes' values.
  */
-function consumeAttributeMatches( viewElement: ViewElement, { consumable }: UpcastConversionApi, matcher: Matcher ): Array<MatchResult> {
-	const matches = matcher.matchAll( viewElement ) || [];
-	const consumedMatches = [];
-
-	for ( const match of matches ) {
-		removeConsumedAttributes( consumable, viewElement, match );
-
-		// We only want to consume attributes, so element can be still processed by other converters.
-		delete match.match.name;
-
-		consumable.consume( viewElement, match.match );
-		consumedMatches.push( match );
+function normalizeGHSAttribute(
+	viewElement: ViewElement,
+	{ attributes, classes, styles }: {
+		attributes: Array<string>;
+		classes: Array<string>;
+		styles: Array<string>;
+	}
+): GHSViewAttributes | null {
+	if ( !attributes.length && !classes.length && !styles.length ) {
+		return null;
 	}
 
-	return consumedMatches;
-}
-
-/**
- * Removes attributes from the given match that were already consumed by other converters.
- */
-function removeConsumedAttributes( consumable: ViewConsumable, viewElement: ViewElement, match: MatchResult ) {
-	for ( const key of [ 'attributes', 'classes', 'styles' ] as const ) {
-		const attributes = match.match[ key ];
+	return {
+		...( attributes.length && {
+			attributes: normalizeAttributes( viewElement, attributes )
+		} ),
 
-		if ( !attributes ) {
-			continue;
-		}
+		...( styles.length && {
+			styles: normalizeStyles( viewElement, styles )
+		} ),
 
-		// Iterating over a copy of an array so removing items doesn't influence iteration.
-		for ( const value of Array.from( attributes ) ) {
-			if ( !consumable.test( viewElement, ( { [ key ]: [ value ] } ) ) ) {
-				removeItemFromArray( attributes, value );
-			}
-		}
-	}
+		...( classes.length && {
+			classes
+		} )
+	};
 }
 
 /**
- * Merges the result of {@link module:engine/view/matcher~Matcher#matchAll} method.
- *
- * @param matches
- * @returns Object with following properties:
- * - attributes Set with matched attribute names.
- * - styles Set with matched style names.
- * - classes Set with matched class names.
+ * Returns list of attributes as a normalized object.
  */
-function mergeMatchResults( matches: Array<MatchResult> ):
-{
-	attributes: Set<string>;
-	styles: Set<string>;
-	classes: Set<string>;
-} {
-	const matchResult = {
-		attributes: new Set<string>(),
-		classes: new Set<string>(),
-		styles: new Set<string>()
-	};
+function normalizeAttributes( viewElement: ViewElement, attributes: Iterable<string> ): Record<string, string> {
```

Comment:
If the function returns a list we can add a prefix `get` to the name of function, WDYT?

So the name would be `getNormalizedAttributes`.

Reply:
Changed to `getAttributes()`.